_posts/2017-10-27-conventional-interfaces.md

```diff
@@ -0,0 +1,225 @@
+
+### Breaking down the implementation of both functions
+
+It was a truly insightful moment when I realized that indeed both functions are almost identical in structure as opposed to how it seemed at first. The reason why the given implementations are so different is the mixing of concerns that hides the signal-flow structure. To prove this, let's take a deeper look at the implementations to analyze where the mixing of concerns happens.
+
+`sumOddSquares`:
+
+* The *enumeration* is partially implemented by the `Leaf` pattern match, and partially by the double recursion, tree-like, in the `Node` pattern match.
+* The *filtering* is mixed with the mapping in the `odd` case in the `Leaf` pattern match.
+* The *folding* or reduction is partially implemented by the `+` that joins the double recursion in the `Node` pattern match, and partially by the `otherwise` case in the `Leaf` pattern match.
+
+Each pattern matching is mixing several concerns, increasing the complexity of the implementation and hiding the signal-flow structure of the computation.
+
+`evenFibs`:
+
+* The *enumeration* is partially implemented by the `k > n` case and party by the `next` expression.
+* The *mapping*, although extracted to the `f` expression, is mixed with the `even` filtering in the `otherwise` case.
+* The *folding* or reduction is partially implemented by `:`, the list constructor, and partially by the `[ ]`, empty list, in the `k > n` case.
+
+As in `sumOddSquares`, each pattern matching in `evenFibs` is mixing different concerns, failing to exhibit the signal-flow structure of the computation.
+
+Let's refactor both functions following the steps defined previously, so that we can expose the [hidden similiarities](#similarities) and the signal-flow structure.
+
+It is worth noting that `.` is the operator for function composition in Haskell.
+
+`sumOddSquares`:
+
+1. *Enumerates* the leaves of the input tree.
+2. *Filters* out the leaves that contain even numbers.
+3. *Maps* the remaining leaves squaring each one.
+4. *Folds* (or reduces) the mapped leaves using `+`, starting at 0.
+
+```
+sumOddSquares :: (Integral a) => BinaryTree a -> a
+sumOddSquares tree = foldr (+) 0 . fmap (^2) . filter odd (enumerateLeaves tree)
+
+enumerateLeaves :: BinaryTree a -> [a]
+enumerateLeaves tree = foldr (:) [] tree
+```
+
+`evenFibs`:
+
+1. *Enumerates* the interval 0 to n.
+2. *Maps* each number to its corresponding Fib.
+3. *Filter* out the odd Fibs.
+4. *Folds* (or reduces) the remaining elements using `:` (List constructor), starting with `[]` (empty list).
+
+```
+evenFibs' :: Integer -> [Integer]
+evenFibs' to = foldr (:) [] . filter even . fmap fib (enumerateInterval 0 to)
+
+enumerateInterval :: Integer  -> Integer -> [Integer]
+enumerateInterval from to = [from..to]
+```
```

Comment:
Can you repeat here the original implementation so we can compare them easily?

Reply:
Done.